plutus-metatheory/test/TestNEAT.hs

```diff
@@ -1,105 +1,260 @@
-  ty  <- withExceptT GenErrorP tyQ
-  tyDB <- withExceptT FVErrorP $ deBruijnTy ty
-  tyN <- withExceptT AgdaErrorP $ case normalizeTypeAgda (AlexPn 0 0 0 <$ tyDB) of
-    Just tyN -> return tyN
-    Nothing  -> throwError ()
-  withExceptT AgdaErrorP $ case checkKindAgda (AlexPn 0 0 0 <$ tyN) (deBruijnifyK (convK k)) of
+prop_runCK :: (Kind (), ClosedTypeG) -> ClosedTermG -> ExceptT TestFail Quote ()
+prop_runCK (k , tyG) tmG = do
+  tm <- withExceptT GenError $ convertClosedTerm tynames names tyG tmG
+  tmDB <- withExceptT FVErrorP $ deBruijnTerm tm
+  case runCKAgda (AlexPn 0 0 0 <$ tmDB) of
     Just _  -> return ()
-    Nothing -> throwError ()
+    Nothing -> throwCtrex (CtrexTermEvaluationFail tyG tmG)
 
--- the agda implementation throws names away, so I guess we need to compare deBruijn terms
-prop_normalizeTypeSound :: TyProp
-prop_normalizeTypeSound k tyG tyQ = isSafe $ do
-  ty <- withExceptT GenErrorP tyQ
-  tyDB <- withExceptT FVErrorP $ deBruijnTy ty
-  tyN1 <- withExceptT AgdaErrorP $ case normalizeTypeAgda (AlexPn 0 0 0 <$ tyDB) of
-    Just tyN -> return tyN
-    Nothing  -> throwError ()
-  ty1 <- withExceptT FVErrorP $ unDeBruijnTy tyN1
+prop_runTCK :: (Kind (), ClosedTypeG) -> ClosedTermG -> ExceptT TestFail Quote ()
+prop_runTCK (k , tyG) tmG = do
+  tm <- withExceptT GenError $ convertClosedTerm tynames names tyG tmG
+  tmDB <- withExceptT FVErrorP $ deBruijnTerm tm
+  case runTCKAgda (AlexPn 0 0 0 <$ tmDB) of
+    Just _  -> return ()
+    Nothing -> throwCtrex (CtrexTermEvaluationFail tyG tmG)
 
-  ty2 <- withExceptT GenErrorP $ toClosedType k (normalizeTypeG tyG)
-  return (ty1 == (AlexPn 0 0 0 <$ ty2))
+prop_run_plcCK_vs_CK :: (Kind (), ClosedTypeG) -> ClosedTermG -> ExceptT TestFail Quote ()
+prop_run_plcCK_vs_CK (k , tyG) tmG = do
+  tm <- withExceptT GenError $ convertClosedTerm tynames names tyG tmG
+  tmPlcCK <- withExceptT CkP $ liftEither $ evaluateCk mempty tm
+  tmDB <- withExceptT FVErrorP $ deBruijnTerm tm
+  tmCK <- withExceptT Ctrex $
+    case runCKAgda (AlexPn 0 0 0 <$ tmDB) of
+      Just tmCK -> return tmCK
+      Nothing   -> throwError (CtrexTermEvaluationFail tyG tmG)
+  tmCKN <- withExceptT FVErrorP $ unDeBruijnTerm tmCK
+  unless (tmPlcCK == (() <$ tmCKN)) $ throwCtrex (CtrexTermEvaluationMismatch tyG tmG tmPlcCK (() <$ tmCKN))
 
-prop_normalizeTypeSame :: TyProp
-prop_normalizeTypeSame k tyG tyQ = isSafe $ do
-  ty <- withExceptT GenErrorP tyQ
-  tyDB <- withExceptT FVErrorP $ deBruijnTy ty
-  tyN1 <- withExceptT AgdaErrorP $ case normalizeTypeAgda (AlexPn 0 0 0 <$ tyDB) of
-    Just tyN -> return tyN
-    Nothing  -> throwError ()
-  ty1 <- withExceptT FVErrorP $ unDeBruijnTy tyN1
+prop_run_L_vs_CK :: (Kind (), ClosedTypeG) -> ClosedTermG -> ExceptT TestFail Quote ()
+prop_run_L_vs_CK (k , tyG) tmG = do
+  tm <- withExceptT GenError $ convertClosedTerm tynames names tyG tmG
+  tmDB <- withExceptT FVErrorP $ deBruijnTerm tm
+  tmCK <- withExceptT Ctrex $
```

Comment:
Quoting my comments on a previous PR:

> Given that we only generate a few thousands of well-kinded types, can we just materialize them into a list using [`search`](http://hackage.haskell.org/package/lazy-search-0.1.2.1/docs/Control-Search.html#v:search) and then manipulate that list directly and use the regular `Test.Tasty` stuff for creating test cases? Both `search` and `ctrex'` call `searchRaw'` under the hood, so this shouldn't cause us problems.

Regenerating everything from scratch in each test sounds inefficient.

Reply:
I now have only two tests, one which runs a bunch of stuff on a generated type and one which runs a bunch of stuff on a generated term.

Reply:
@effectfully how would this work with `search` and `Test.tasty`?

```
search :: (Enumerable a, Coolean cool) => Int -> (a -> cool) -> IO [a]
```

Would you run `search` with a predicate that guarantees things are well kinded/typed and then perform a unit test on the list such as 'normalizing the type preserves the kind' for all elements of the list using `Test.Tasty.HUnit`?

Reply:
Yep. I'd perhaps make a unit test for each single element of the generated list and then compose all those tests together using `Test.Tasty.testGroup`. But not entirely sure, maybe having a single unit test is more efficient or convenient or something.